File: sources/dev/authentication/src/handlers/oauth2.rs
use axum::{extract::State, Json};
use sea_orm::{ColumnTrait, EntityTrait, QueryFilter};
use serde::{Deserialize, Serialize};

use crate::auth::middleware::AuthenticatedApp;
use crate::auth::oauth2 as oauth2_util;
use crate::auth::password::verify_password;
use crate::error::AppError;
use crate::AppState;

// --- Request / Response types ---

#[derive(Debug, Deserialize)]
pub struct TokenRequest {
    pub grant_type: String,
    // authorization_code flow
    pub code: Option<String>,
    pub redirect_uri: Option<String>,
    pub code_verifier: Option<String>,
    // password flow
    pub username: Option<String>,
    pub password: Option<String>,
    // refresh_token flow
    pub refresh_token: Option<String>,
    // common
    pub scope: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct OAuthTokenResponse {
    pub access_token: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub refresh_token: Option<String>,
    pub token_type: String,
    pub expires_in: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct RevokeRequest {
    pub token: String,
}

#[derive(Debug, Deserialize)]
pub struct IntrospectRequest {
    pub token: String,
}

#[derive(Debug, Serialize)]
pub struct IntrospectResponse {
    pub active: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sub: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aud: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exp: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope: Option<String>,
}

// --- Handlers ---

pub async fn token(
    auth_app: AuthenticatedApp,
    State(state): State<AppState>,
    Json(req): Json<TokenRequest>,
) -> Result<Json<OAuthTokenResponse>, AppError> {
    match req.grant_type.as_str() {
        "authorization_code" => handle_authorization_code(&state, &auth_app, &req).await,
        "client_credentials" => handle_client_credentials(&state, &auth_app).await,
        "refresh_token" => handle_refresh_token(&state, &auth_app, &req).await,
        "password" => handle_password_grant(&state, &auth_app, &req).await,
        _ => Err(AppError::BadRequest(format!(
            "Unsupported grant_type: {}",
            req.grant_type
        ))),
    }
}

async fn handle_authorization_code(
    state: &AppState,
    auth_app: &AuthenticatedApp,
    req: &TokenRequest,
) -> Result<Json<OAuthTokenResponse>, AppError> {
    let code = req.code.as_deref().ok_or(AppError::BadRequest(
        "Missing 'code' parameter".to_string(),
    ))?;
    let redirect_uri = req.redirect_uri.as_deref().ok_or(AppError::BadRequest(
        "Missing 'redirect_uri' parameter".to_string(),
    ))?;

    let (user_id, scopes) = oauth2_util::exchange_auth_code(
        &state.db,
        code,
        &auth_app.app_id,
        redirect_uri,
        req.code_verifier.as_deref(),
    )
    .await?;

    // Look up user for role
    let user = entity::user::Entity::find_by_id(&user_id)
        .one(&state.db)
        .await?
        .ok_or(AppError::UserNotFound)?;

    if !user.is_active {
        return Err(AppError::Forbidden);
    }

    let access_token = state.jwt.issue_access_token(&user_id, &auth_app.client_id, scopes.clone(), &user.role)?;
    let refresh_token = oauth2_util::generate_refresh_token();

    oauth2_util::store_refresh_token(
        &state.db,
        &user_id,
        &auth_app.app_id,
        &refresh_token,
        &scopes,
        None,
        state.config.jwt_refresh_token_expiry_days,
    )
    .await?;

    Ok(Json(OAuthTokenResponse {
        access_token,
        refresh_token: Some(refresh_token),
        token_type: "Bearer".to_string(),
        expires_in: state.config.jwt_access_token_expiry_secs,
        scope: Some(scopes.join(" ")),
    }))
}

async fn handle_client_credentials(
    state: &AppState,
    auth_app: &AuthenticatedApp,
) -> Result<Json<OAuthTokenResponse>, AppError> {
    let access_token = state.jwt.issue_app_token(&auth_app.app_id)?;

    Ok(Json(OAuthTokenResponse {
        access_token,
        refresh_token: None,
        token_type: "Bearer".to_string(),
        expires_in: state.config.jwt_access_token_expiry_secs,
        scope: None,
    }))
}

async fn handle_refresh_token(
    state: &AppState,
    auth_app: &AuthenticatedApp,
    req: &TokenRequest,
) -> Result<Json<OAuthTokenResponse>, AppError> {
    let refresh_token_str = req.refresh_token.as_deref().ok_or(AppError::BadRequest(
        "Missing 'refresh_token' parameter".to_string(),
    ))?;

    let (user_id, new_refresh_token, scopes) = oauth2_util::rotate_refresh_token(
        &state.db,
        refresh_token_str,
        &auth_app.app_id,
        state.config.jwt_refresh_token_expiry_days,
    )
    .await?;

    // Look up user for role
    let user = entity::user::Entity::find_by_id(&user_id)
        .one(&state.db)
        .await?
        .ok_or(AppError::UserNotFound)?;

    if !user.is_active {
        return Err(AppError::Forbidden);
    }

    let access_token = state.jwt.issue_access_token(&user_id, &auth_app.client_id, scopes.clone(), &user.role)?;

    Ok(Json(OAuthTokenResponse {
        access_token,
        refresh_token: Some(new_refresh_token),
        token_type: "Bearer".to_string(),
        expires_in: state.config.jwt_access_token_expiry_secs,
        scope: Some(scopes.join(" ")),
    }))
}

async fn handle_password_grant(
    state: &AppState,
    auth_app: &AuthenticatedApp,
    req: &TokenRequest,
) -> Result<Json<OAuthTokenResponse>, AppError> {
    let username = req.username.as_deref().ok_or(AppError::BadRequest(
        "Missing 'username' parameter".to_string(),
    ))?;
    let password = req.password.as_deref().ok_or(AppError::BadRequest(
        "Missing 'password' parameter".to_string(),
    ))?;

    // Find user by email
    let user = entity::user::Entity::find()
        .filter(entity::user::Column::Email.eq(username))
        .one(&state.db)
        .await?
        .ok_or(AppError::InvalidCredentials)?;

    // Find password account
    let account = entity::account::Entity::find()
        .filter(entity::account::Column::UserId.eq(&user.id))
        .filter(entity::account::Column::ProviderId.eq("password"))
        .one(&state.db)
        .await?
        .ok_or(AppError::InvalidCredentials)?;

    let credential = account.credential.ok_or(AppError::InvalidCredentials)?;
    if !verify_password(password, &credential)? {
        return Err(AppError::InvalidCredentials);
    }

    // Determine scopes
    let app = entity::application::Entity::find_by_id(&auth_app.app_id)
        .one(&state.db)
        .await?
        .ok_or(AppError::ApplicationNotFound)?;

    let allowed_scopes: Vec<String> =
        serde_json::from_str(&app.allowed_scopes).unwrap_or_default();

    let scopes = if let Some(ref scope_str) = req.scope {
        let requested: Vec<String> = scope_str.split(' ').map(|s| s.to_string()).collect();
        // Filter to only allowed scopes
        requested
            .into_iter()
            .filter(|s| allowed_scopes.contains(s))
            .collect()
    } else {
        allowed_scopes
    };

    if !user.is_active {
        return Err(AppError::Forbidden);
    }

    let access_token = state.jwt.issue_access_token(&user.id, &auth_app.client_id, scopes.clone(), &user.role)?;
    let refresh_token = oauth2_util::generate_refresh_token();

    oauth2_util::store_refresh_token(
        &state.db,
        &user.id,
        &auth_app.app_id,
        &refresh_token,
        &scopes,
        None,
        state.config.jwt_refresh_token_expiry_days,
    )
    .await?;

    Ok(Json(OAuthTokenResponse {
        access_token,
        refresh_token: Some(refresh_token),
        token_type: "Bearer".to_string(),
        expires_in: state.config.jwt_access_token_expiry_secs,
        scope: Some(scopes.join(" ")),
    }))
}

pub async fn revoke(
    _auth_app: AuthenticatedApp,
    State(state): State<AppState>,
    Json(req): Json<RevokeRequest>,
) -> Result<Json<serde_json::Value>, AppError> {
    // Try to revoke as refresh token
    let _ = oauth2_util::revoke_refresh_token(&state.db, &req.token).await;
    // Per RFC 7009, always return 200
    Ok(Json(serde_json::json!({})))
}

pub async fn introspect(
    _auth_app: AuthenticatedApp,
    State(state): State<AppState>,
    Json(req): Json<IntrospectRequest>,
) -> Result<Json<IntrospectResponse>, AppError> {
    // Try to verify as access token
    match state.jwt.verify_access_token(&req.token) {
        Ok(claims) => Ok(Json(IntrospectResponse {
            active: true,
            sub: Some(claims.sub),
            aud: Some(claims.aud),
            exp: Some(claims.exp),
            scope: Some(claims.scopes.join(" ")),
        })),
        Err(_) => Ok(Json(IntrospectResponse {
            active: false,
            sub: None,
            aud: None,
            exp: None,
            scope: None,
        })),
    }
}

File: sources/dev/authentication-go/internal/handlers/oauth2.go
package handlers

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/zhaochy1990/auth-service/internal/apperror"
	"github.com/zhaochy1990/auth-service/internal/auth"
	"github.com/zhaochy1990/auth-service/internal/middleware"
)

// --- Request / Response types ---

type tokenRequest struct {
	GrantType string `json:"grant_type"`
	// authorization_code flow
	Code         *string `json:"code"`
	RedirectURI  *string `json:"redirect_uri"`
	CodeVerifier *string `json:"code_verifier"`
	// password flow
	Username *string `json:"username"`
	Password *string `json:"password"`
	// refresh_token flow
	RefreshToken *string `json:"refresh_token"`
	// common
	Scope *string `json:"scope"`
}

type oauthTokenResponse struct {
	AccessToken  string  `json:"access_token"`
	RefreshToken *string `json:"refresh_token,omitempty"`
	TokenType    string  `json:"token_type"`
	ExpiresIn    int64   `json:"expires_in"`
	Scope        *string `json:"scope,omitempty"`
}

type revokeRequest struct {
	Token string `json:"token"`
}

type introspectRequest struct {
	Token string `json:"token"`
}

type introspectResponse struct {
	Active bool    `json:"active"`
	Sub    *string `json:"sub,omitempty"`
	Aud    *string `json:"aud,omitempty"`
	Exp    *int64  `json:"exp,omitempty"`
	Scope  *string `json:"scope,omitempty"`
}

// --- Handlers ---

// Token implements the OAuth2 token endpoint (multiple grant types).
func (h *Handler) Token(c *gin.Context) {
	var req tokenRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		middleware.RespondError(c, apperror.BadRequest("Invalid request body"))
		return
	}
	switch req.GrantType {
	case "authorization_code":
		h.handleAuthorizationCode(c, &req)
	case "client_credentials":
		h.handleClientCredentials(c)
	case "refresh_token":
		h.handleRefreshTokenGrant(c, &req)
	case "password":
		h.handlePasswordGrant(c, &req)
	default:
		middleware.RespondError(c, apperror.BadRequest("Unsupported grant_type: "+req.GrantType))
	}
}

func (h *Handler) handleAuthorizationCode(c *gin.Context, req *tokenRequest) {
	ctx := c.Request.Context()
	if req.Code == nil {
		middleware.RespondError(c, apperror.BadRequest("Missing 'code' parameter"))
		return
	}
	if req.RedirectURI == nil {
		middleware.RespondError(c, apperror.BadRequest("Missing 'redirect_uri' parameter"))
		return
	}
	userID, scopes, err := auth.ExchangeAuthCode(ctx, h.Repo, *req.Code, middleware.AppID(c), *req.RedirectURI, req.CodeVerifier)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	user, err := h.Repo.Users().FindByID(ctx, userID)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if user == nil {
		middleware.RespondError(c, apperror.UserNotFound())
		return
	}
	if !user.IsActive {
		middleware.RespondError(c, apperror.Forbidden())
		return
	}
	membership := h.resolveMembership(ctx, user)
	accessToken, err := h.JWT.IssueAccessToken(userID, middleware.ClientID(c), scopes, user.Role, membership, user.Name)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	refreshToken := auth.GenerateRefreshToken()
	if err := auth.StoreRefreshToken(ctx, h.Repo, userID, middleware.AppID(c), refreshToken, scopes, nil, h.Cfg.JWTRefreshTokenExpiryDays); err != nil {
		middleware.RespondError(c, err)
		return
	}
	scopeStr := strings.Join(scopes, " ")
	c.JSON(http.StatusOK, oauthTokenResponse{
		AccessToken:  accessToken,
		RefreshToken: strPtr(refreshToken),
		TokenType:    "Bearer",
		ExpiresIn:    h.Cfg.JWTAccessTokenExpirySecs,
		Scope:        &scopeStr,
	})
}

func (h *Handler) handleClientCredentials(c *gin.Context) {
	accessToken, err := h.JWT.IssueAppToken(middleware.AppID(c))
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	c.JSON(http.StatusOK, oauthTokenResponse{
		AccessToken: accessToken,
		TokenType:   "Bearer",
		ExpiresIn:   h.Cfg.JWTAccessTokenExpirySecs,
	})
}

func (h *Handler) handleRefreshTokenGrant(c *gin.Context, req *tokenRequest) {
	ctx := c.Request.Context()
	if req.RefreshToken == nil {
		middleware.RespondError(c, apperror.BadRequest("Missing 'refresh_token' parameter"))
		return
	}
	userID, newRefreshToken, scopes, err := auth.RotateRefreshToken(ctx, h.Repo, *req.RefreshToken, middleware.AppID(c), h.Cfg.JWTRefreshTokenExpiryDays)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	user, err := h.Repo.Users().FindByID(ctx, userID)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if user == nil {
		middleware.RespondError(c, apperror.UserNotFound())
		return
	}
	if !user.IsActive {
		middleware.RespondError(c, apperror.Forbidden())
		return
	}
	membership := h.resolveMembership(ctx, user)
	accessToken, err := h.JWT.IssueAccessToken(userID, middleware.ClientID(c), scopes, user.Role, membership, user.Name)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	scopeStr := strings.Join(scopes, " ")
	c.JSON(http.StatusOK, oauthTokenResponse{
		AccessToken:  accessToken,
		RefreshToken: strPtr(newRefreshToken),
		TokenType:    "Bearer",
		ExpiresIn:    h.Cfg.JWTAccessTokenExpirySecs,
		Scope:        &scopeStr,
	})
}

func (h *Handler) handlePasswordGrant(c *gin.Context, req *tokenRequest) {
	ctx := c.Request.Context()
	if req.Username == nil {
		middleware.RespondError(c, apperror.BadRequest("Missing 'username' parameter"))
		return
	}
	if req.Password == nil {
		middleware.RespondError(c, apperror.BadRequest("Missing 'password' parameter"))
		return
	}
	user, err := h.Repo.Users().FindByEmail(ctx, *req.Username)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if user == nil {
		middleware.RespondError(c, apperror.InvalidCredentials())
		return
	}
	account, err := h.Repo.Accounts().FindByUserAndProvider(ctx, user.ID, "password")
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if account == nil || account.Credential == nil {
		middleware.RespondError(c, apperror.InvalidCredentials())
		return
	}
	ok, err := auth.VerifyPassword(*req.Password, *account.Credential)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if !ok {
		middleware.RespondError(c, apperror.InvalidCredentials())
		return
	}

	app, err := h.Repo.Applications().FindByID(ctx, middleware.AppID(c))
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	if app == nil {
		middleware.RespondError(c, apperror.ApplicationNotFound())
		return
	}
	allowedScopes := auth.DecodeStringArray(app.AllowedScopes)

	var scopes []string
	if req.Scope != nil {
		for _, s := range strings.Split(*req.Scope, " ") {
			if contains(allowedScopes, s) {
				scopes = append(scopes, s)
			}
		}
	} else {
		scopes = allowedScopes
	}

	if !user.IsActive {
		middleware.RespondError(c, apperror.Forbidden())
		return
	}
	membership := h.resolveMembership(ctx, user)
	accessToken, err := h.JWT.IssueAccessToken(user.ID, middleware.ClientID(c), scopes, user.Role, membership, user.Name)
	if err != nil {
		middleware.RespondError(c, err)
		return
	}
	refreshToken := auth.GenerateRefreshToken()
	if err := auth.StoreRefreshToken(ctx, h.Repo, user.ID, middleware.AppID(c), refreshToken, scopes, nil, h.Cfg.JWTRefreshTokenExpiryDays); err != nil {
		middleware.RespondError(c, err)
		return
	}
	scopeStr := strings.Join(scopes, " ")
	c.JSON(http.StatusOK, oauthTokenResponse{
		AccessToken:  accessToken,
		RefreshToken: strPtr(refreshToken),
		TokenType:    "Bearer",
		ExpiresIn:    h.Cfg.JWTAccessTokenExpirySecs,
		Scope:        &scopeStr,
	})
}

// Revoke revokes a refresh token. Per RFC 7009, always returns 200.
func (h *Handler) Revoke(c *gin.Context) {
	var req revokeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		middleware.RespondError(c, apperror.BadRequest("Invalid request body"))
		return
	}
	_ = auth.RevokeRefreshToken(c.Request.Context(), h.Repo, req.Token)
	c.JSON(http.StatusOK, gin.H{})
}

// Introspect reports whether an access token is active (RFC 7662 subset).
func (h *Handler) Introspect(c *gin.Context) {
	var req introspectRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		middleware.RespondError(c, apperror.BadRequest("Invalid request body"))
		return
	}
	claims, err := h.JWT.VerifyAccessToken(req.Token)
	if err != nil {
		c.JSON(http.StatusOK, introspectResponse{Active: false})
		return
	}
	scope := strings.Join(claims.Scopes, " ")
	exp := claims.Exp
	c.JSON(http.StatusOK, introspectResponse{
		Active: true,
		Sub:    strPtr(claims.Sub),
		Aud:    strPtr(claims.Aud),
		Exp:    &exp,
		Scope:  &scope,
	})
}

func contains(ss []string, s string) bool {
	for _, v := range ss {
		if v == s {
			return true
		}
	}
	return false
}
